### ISSUE TITLE
Provide demos for conditional generation

### ISSUE THREAD

**Issue by bugsz** [atong01/conditional-flow-matching#30]
Hi, I noticed that you have provided a demo on unconditional generation on the MNIST dataset. Do you plan to provide some demos on **conditional generation**? That would be of great help to me.
Thanks again for your valuable work and codebase!

**Comment by gevmin94** [comment]
Hi, I extended provided MNIST example with a specific number conditional generation.

First at inference time haven't a clear idea of how to pass additional conditional inputs in our case digit labels using `NerualODE` from `torchdyn`. Instead, I used  `odeint` function from `torchdiffeq` like this
```
 l = torch.tensor([digit_label]).to(device)
 out = torchdiffeq.odeint(
          lambda t, x: model.forward(t, x, l),
          torch.randn(1,1,28,28).to(device), 
          torch.linspace(0, 1, 2).to(device),
          atol=1e-4, 
          rtol=1e-4,
          method='dopri5'
  )   
```
In a recent, [Voicebox](https://arxiv.org/pdf/2306.15687.pdf) article from Meta also written they used `torchdiffeq`.
Without the conditional case `odeint` function  I got similar results.  However, I noticed solver doesn't end at exactly 1.0 time in t_span in `torchdiffeq` case instead it exceeds time slightly 1.1, whent I give t_span=[0.0, 2.0] it exceeds even larger 11.1. 

When I tried different label conditioning settings, the model did not generate numbers of the given class label but produced random numbers similar to the unconditional case. In the original Unet implementation, I see there is an option of conditioning label embeddings to sum timestamp embeddings. I tried concat instead of summing, and also I tried to condition label embedding by summing noise input changing embedding dimension 28*28. Play with different hyper params like learning rate, increased number of residual blocks, model dim, and epochs. All tried settings model behaves like unconditional case generating random numbers ignoring class label. 
I wonder if I missed something crucial or if there's a need to modify the loss computation or any other aspect of the model.

**Comment by bugsz** [comment]
Thanks for your detailed explanation! I will look into it and see if I can find something new!

**Comment by atong01** [comment]
@gevmin94 This is very interesting! Are you willing to share this code? Would love to take a look. I'm busy with NeurIPS rebuttals through this week but will definitely be looking into this next week. 

The model training needs to be modified a bit for the conditional generation to work, but this should not be too challenging. 

I'm not super surprised the solver slightly exceeds the time specified because of the adaptive solver used, but the 11.1 seems quite a bit larger than I would expect.

**Comment by gevmin94** [comment]
I have opened [PR]( https://github.com/atong01/conditional-flow-matching/pull/33 ) for my current changes. Additionally, I can provide a summary of the changes below. 

Training Loop:

```python
for epoch in range(n_epochs):
    for i, data in enumerate(train_loader):
        optimizer.zero_grad()
        x1 = data[0].to(device)
        y = data[1].to(device)
        x0 = torch.randn_like(x1)
        t, xt, ut = FM.sample_location_and_conditional_flow(x0, x1)
        vt = model(t, xt, y)
        loss = torch.mean((vt - ut) ** 2)
        loss.backward()
        optimizer.step()
        print(f'epoch: {epoch}, steps: {i}, loss: {loss.item():.4}', end='\r')
```


Inference:

```python
USE_TORCH_DIFFEQ = True
digit_label = 8
l = torch.tensor([digit_label]*100).to(device)
with torch.no_grad():
    if USE_TORCH_DIFFEQ:
        traj = torchdiffeq.odeint(
            lambda t, x: model.forward(t, x, l),
            torch.randn(100,1,28,28).to(device), 
            torch.linspace(0, 1, 2).to(device),
            atol=1e-4, 
            rtol=1e-4,
            method='dopri5'
        )   
    else:
        traj = node.trajectory(
            torch.randn(100, 1, 28, 28).to(device),
            t_span=torch.linspace(0, 1, 2).to(device),
        )
grid = make_grid(
    traj[-1, :100].view([-1, 1, 28, 28]).clip(-1, 1), value_range=(-1, 1), padding=0, nrow=10
)
img = ToPILImage()(grid)
plt.imshow(img)
plt.show()
```

Is there anything else that needs to be modified in the training process besides passing labels to the model?

**Comment by bugsz** [comment]
> I have opened [PR](https://github.com/atong01/conditional-flow-matching/pull/33) for my current changes. Additionally, I can provide a summary of the changes below.
> 
> Training Loop:
> 
> ```python
> for epoch in range(n_epochs):
>     for i, data in enumerate(train_loader):
>         optimizer.zero_grad()
>         x1 = data[0].to(device)
>         y = data[1].to(device)
>         x0 = torch.randn_like(x1)
>         t, xt, ut = FM.sample_location_and_conditional_flow(x0, x1)
>         vt = model(t, xt, y)
>         loss = torch.mean((vt - ut) ** 2)
>         loss.backward()
>         optimizer.step()
>         print(f'epoch: {epoch}, steps: {i}, loss: {loss.item():.4}', end='\r')
> ```
> 
> Inference:
> 
> ```python
> USE_TORCH_DIFFEQ = True
> digit_label = 8
> l = torch.tensor([digit_label]*100).to(device)
> with torch.no_grad():
>     if USE_TORCH_DIFFEQ:
>         traj = torchdiffeq.odeint(
>             lambda t, x: model.forward(t, x, l),
>             torch.randn(100,1,28,28).to(device), 
>             torch.linspace(0, 1, 2).to(device),
>             atol=1e-4, 
>             rtol=1e-4,
>             method='dopri5'
>         )   
>     else:
>         traj = node.trajectory(
>             torch.randn(100, 1, 28, 28).to(device),
>             t_span=torch.linspace(0, 1, 2).to(device),
>         )
> grid = make_grid(
>     traj[-1, :100].view([-1, 1, 28, 28]).clip(-1, 1), value_range=(-1, 1), padding=0, nrow=10
> )
> img = ToPILImage()(grid)
> plt.imshow(img)
> plt.show()
> ```
> 
> Is there anything else that needs to be modified in the training process besides passing labels to the model?

I notice that the result is still not that satisfying, but I'm not sure what is missing here.

Besides, I tried another task, which is to complete the number given some part of the original image is masked. I tried to concatenate condition (masked image) and random noise on the channel dimension and make a linear projection before going into the U-Net, which resembles the Voicebox. Below is the modification.
`
C, H, W = dim
self.ctx_proj = nn.Parameter(torch.randn(H, W, C * 2, C))

if ctx is not None:
    assert x.shape == ctx.shape
    x = torch.cat((x, ctx), dim=1) # [B, 2, 28, 28]
    x = torch.matmul(x.permute(2, 3, 0, 1), self.ctx_proj).permute(2, 3, 0, 1) # [B, 1, 28, 28]
`

However, the output image is still a random one.

**Comment by atong01** [comment]
This is great! @gevmin94, there is one subtle bug in the training loop. I'll make a change to fix this and post here when I have it working. Basically, when sampling the OT sampler scrambles the rows of the `xt`s relative to the `y` variables. We need to scramble the condition in the same way. Note that both of your PRs should work as is for the `ConditionalFlowMatcher` instead of the `ExactOptimalTransportConditionalFlowMatcher`. I'm adding a way to scramble the conditions in the same way and will post here when its done.

I think what @bugsz is trying to do with image generation will also be easy with this modification.

**Comment by bugsz** [comment]
I think I may make a mistake since it doesn't make sense to mix conditions and random noise only before the U-Net. Will try to see if there is anything I can do.

**Comment by atong01** [comment]
See updates to PR #33

**Comment by gevmin94** [comment]
I now understand what was previously unclear. Thank you for providing clarification. Before delving into coding, I should thoroughly read your impressive work. I was initially not aware of the distinction between Cond. OT and Marginal OT. Without having read your work, I mistakenly believed that `OT-CFM` was the same as Lipman's FM algorithm, that's why even not doubting `OT-CFM` changes the order of minibatch items. The terminology surrounding OT was perplexing to me.

Thanks again for your assistance, this saves a lot of time for me.
